block/gcs/adapter.go

```diff
@@ -0,0 +1,438 @@
+package gcs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/treeverse/lakefs/block"
+	"github.com/treeverse/lakefs/logging"
+)
+
+const (
+	BlockstoreType = "gcs"
+
+	DefaultStreamingChunkSize    = 2 << 19         // 1MiB by default per chunk
+	DefaultStreamingChunkTimeout = time.Second * 1 // if we haven't read DefaultStreamingChunkSize by this duration, write whatever we have as a chunk
+)
+
+var (
+	ErrGCS                     = errors.New("gcs error")
+	ErrMissingETag             = fmt.Errorf("%w: missing ETag", ErrGCS)
+	ErrInventoryNotImplemented = errors.New("inventory feature not implemented for gcs storage adapter")
+)
+
+func resolveNamespace(obj block.ObjectPointer) (block.QualifiedKey, error) {
+	qualifiedKey, err := block.ResolveNamespace(obj.StorageNamespace, obj.Identifier)
+	if err != nil {
+		return qualifiedKey, err
+	}
+	if qualifiedKey.StorageType != block.StorageTypeS3 {
```

Comment:
Perhaps we can make it "gs://" instead of "s3://" for the storage namespace of the repo, and then replace it here to s3.
This way the functionality will not be changed on the next "increment"

Possible?

Reply:
possible - we will have 'gs' be the block adapter name with the configuration in the next iteration